feat(timers): track time using a start/stop timer

chadfawcett opened the pull request:
Could someone from @tickbin/core review this?

* `tick start` starts a timer
* `tick stop` stops the timer and commits
* `tick timer` lists details about timer
* `tick cancel-timer` cancels timer without committing

Resolves #191

chrisfosterelli commented on src/commands/tick-cancel-timer.js:
```diff
@@ -0,0 +1,26 @@
+function removeTimer(timersDoc) {
+  const timer = timersDoc.timers.pop()
+
+  if (!timer) throw { message: 'You do not have a timer started' }
```
Instead of throwing an object with a `message` property, it would be much more idiomatic to throw an `Error`. While this works as it is, any future code that calls `removeTimer` may want to call `.trace()` or other builtin functionality on the error.

chrisfosterelli commented on src/commands/tick-start.js:
```diff
@@ -0,0 +1,52 @@
+function saveTimer(timersDoc, originalMessage) {
+  //  For now only allow one timer at a time
+  if (timersDoc.timers.length > 0) {
+    throw { message: 'You already have a timer running. You can run:\n'
```
Same as [above](https://github.com/tickbin/tickbin/pull/270/files#r75430566)

chrisfosterelli commented on src/commands/tick-stop.js:
```diff
@@ -0,0 +1,77 @@
+  .then(timersDoc => parseMessage(timersDoc, argv._[1]))
+  .then(commitTimer)
+  .then(writeSaved)
+  .catch(err => console.log(`\nCould not stop your timer\n${err.message}`))
```
Not sure if there is a reason for this, but why does `stop` print a newline before the error message but `start` [doesn't?](https://github.com/tickbin/tickbin/pull/270/files#diff-ecd55f42665b0b941f00cf33cc9a3888R29).

Likely noone would notice, so mostly just curious if it was deliberate.

chrisfosterelli commented on src/commands/tick-stop.js:
```diff
@@ -0,0 +1,77 @@
+function parseMessage(timersDoc, newMessage) {
+  const timer = timersDoc.timers.pop()
+
+  if (!timer) throw { message: 'You do not have a timer started' }
```
Same as [above](https://github.com/tickbin/tickbin/pull/270/files#r75430566)

chrisfosterelli commented on src/commands/tick-stop.js:
```diff
@@ -0,0 +1,77 @@
+    return new Promise((resolve, reject) => {
+      prompt.get('message', (err, res) => {
+        if (err && err.message === 'canceled')
+          return reject({ message: 'You canceled the stop command' })
```
Rejections should be done instances of `Error` too

chrisfosterelli commented on src/commands/tick-timer.js:
```diff
@@ -0,0 +1,24 @@
+function getTimer(timersDoc) {
+  const timer = timersDoc.timers[0]
+
+  if (!timer) throw { message: 'You do not have a timer started' }
```
`Error` instance needed :)

chrisfosterelli commented on src/commands/tick-timer.js:
```diff
@@ -0,0 +1,24 @@
+  db.get('_local/timers')
+  .then(getTimer)
+  .then(writeTimer)
+  .catch(err => console.log(`Could not list your timer\n${err.message}`))
```
`console.error` is better fit here.

chrisfosterelli commented on src/commands/tick-cancel-timer.js:
```diff
@@ -0,0 +1,26 @@
+  db.get('_local/timers')
+  .then(removeTimer)
+  .then(writeRemoveTimer)
+  .catch(err => console.log(`Could not cancel your timer\n${err.message}`))
```
`console.error` is better fit here.

chrisfosterelli commented:
@chadfawcett I added a couple of best practice nitpicks using inline comments. More general feedback:

1) Is there a way to abstract some of this out? We have _alot_ of logic in some of these commands (particularly `stop`) compared to other commands we have, and it's all untested because it resides in `/commands`. It'd be great if we could pull some of that out into testable helper libraries, but I'm not sure how practical that is given it's pretty IO tied. What do you think?

2) Do we need to call the cancel command `cancel-timer`? It feels out of pattern with `start`, `stop`, and `timer`. I could see how it may be vague without timer context, but so are `start` and `stop` so having `cancel` isn't too much different.

chrisfosterelli commented on src/commands/tick-start.js:
```diff
@@ -0,0 +1,52 @@
+    //  If the timers document exists, pass it to saveTimer. Otherwise pass the
+    //  defTimersDoc
+    timerDoc => saveTimer(timerDoc, argv._[1]),
+    () => saveTimer(defTimersDoc, argv._[1])
```
There are cases other than the doc missing where this promise might be rejected. Maybe it's worth checking that `err.name == 'not_found'` to ensure that was the case first? Also, not sure if it was deliberate to use the second parameter to the `then` but you can repeat catches if you want:

```
db.get('_loca/timers')
.then(/* Handle existing timer */)
.catch(/* Handle missing timer, rethrow if it was some other error */)
.then(writeSavedTimer)
.catch(err => console.log('Could not start timer'...))
```

chadfawcett commented:
@chrisfosterelli Thanks for the review!

* For the throwing of the object with a message field, I was thinking of using an Error but thought it might not work with what I was doing at first. I should have checked.
* As for the newline in the stop error. There was a case where it would output the error on the same line as other output causing it to look weird. I can't seem to replicate that so in favour of consistency I've removed the newline character.
* I think I knew I could repeat the catches at the end of the chain but didn't know I could put them in the middle. It makes sense though. Thanks for the tip!
* My thought behind the `cancel-timer` name was that like you said, without timer context it is a bit vague. Start and stop, in the context of 'time', implied to me that it would be a timer/stopwatch type behaviour. However, the `stop` command to me didn't necessarily imply a timer (I don't know why now that I think about it). I've changed it to just be `tick stop` in favour of the consistency and ease of typing.

I'll take a look at abstracting some of the logic out in order to test it.
